backend/cfg/cfgize.ml

```diff
@@ -787,22 +787,79 @@ let fundecl :
     State.make ~fun_name ~tailrec_label ~contains_calls:fun_contains_calls
       cfg.blocks
   in
-  (* CR run [add_blocks] here but only for Iname_for_debugger instructions *)
-  State.add_block state ~label:(Cfg.entry_label cfg)
+  (* Ensure any name-for-debugger operations come at the very start, even before
+     the prologue. Otherwise, when the debugger is standing at the start of a
+     function, parameters may be unavailable. *)
+  let rev_name_for_debugger_ops_before_prologue, fun_body =
+    let rec collect (instr : Mach.instruction) ops =
+      match[@ocaml.warning "-4"] instr.desc with
+      | Iop
+          (Iname_for_debugger
+            { ident; which_parameter; provenance; is_assignment; regs }) ->
+        let cfg_op : Cfg.operation =
+          Name_for_debugger
+            { ident; which_parameter; provenance; is_assignment; regs }
+        in
+        (* We use the ordering of the collected instructions to identify the
+           last name-for-debugger operation, whose availability set we will use
+           for the inserted Cfg block, and crucially also to copy over to the
+           prologue. The last step is important so that the parameters' ranges
+           start at the very beginning of the function's code. *)
+        collect instr.next ((instr, cfg_op) :: ops)
+      | _ -> ops, instr
+    in
+    collect fun_body []
+  in
+  let next_label, instr_to_copy =
+    match rev_name_for_debugger_ops_before_prologue with
+    | [] -> Cfg.entry_label cfg, fun_body
+    | _ :: _ ->
+      let instrs =
+        List.map
+          (fun ((naming_op_instr : Mach.instruction), naming_cfg_op) ->
+            let instr = make_instruction state ~desc:(Cfg.Op naming_cfg_op) in
+            instr.dbg <- naming_op_instr.dbg;
+            instr.fdo <- Fdo_info.none;
+            instr)
+          (List.rev rev_name_for_debugger_ops_before_prologue)
+      in
+      let next_label = Cmm.new_label () in
+      let last_naming_op_instr =
+        fst (List.hd rev_name_for_debugger_ops_before_prologue)
+      in
+      State.add_block state ~label:(Cfg.entry_label cfg)
+        ~block:
+          { start = Cfg.entry_label cfg;
+            body = DLL.of_list instrs;
+            terminator =
+              copy_instruction_no_reg state last_naming_op_instr
+                ~desc:(Cfg.Always next_label);
+            (* See [Cfg.register_predecessors_for_all_blocks] *)
+            predecessors = Label.Set.empty;
+            stack_offset = invalid_stack_offset;
+            exn = None;
+            can_raise = false;
+            is_trap_handler = false;
+            dead = false;
+            cold = false
+          };
+      next_label, last_naming_op_instr
```

Comment:
I am not sure if this should be last_naming_op.next ?

Reply:
No, I don't think so - we need to ensure that the availability sets from the last name-for-debugger operation are replicated onto the prologue and tailrec entry point instructions.  Otherwise the prologue would end up with e.g. an empty available-before set, so the parameters wouldn't be visible in the debugger at the top of the function.

There might be some question about whether the `Debuginfo.t` from the next instruction here is better, but I don't think this matters - the emitter already emits `fun_dbg` at the top of the function.